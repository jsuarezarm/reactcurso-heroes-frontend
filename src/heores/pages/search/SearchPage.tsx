import { CustomJumbotron } from '@/components/custom/CustomJumbotron';
import { HeroStats } from '@/heores/componentes/HeroStats';
import { SearchControls } from './ui/SearchControls';
import { CustomBreadcrumbs } from '@/components/custom/CustomBreadcrumbs';
import { HeroGrid } from '@/heores/componentes/HeroGrid';
import { useSearchParams } from 'react-router';
import { useSearch } from '@/heores/hooks/useSearch';

export const SearchPage = () => {

  const [searchParams] = useSearchParams();
  const name = searchParams.get('name') ?? undefined;
  const strength = searchParams.get('strength') ?? undefined;

  const { data: searchHeroes = [] } = useSearch({ name, strength });

  return (
    <>
      {/* Header */}
      <CustomJumbotron
        title="Búsqueda de Superhéores"
        description="Descubre, explora y administra superhéroes y villanos"
      />

      <CustomBreadcrumbs />

      {/* Stats Dashboard */}
      <HeroStats />

      {/* Filter and search */}
      <SearchControls />

      {/* Heroes */}
      <HeroGrid heroes={searchHeroes} />
    </>
  );
};

export default SearchPage;