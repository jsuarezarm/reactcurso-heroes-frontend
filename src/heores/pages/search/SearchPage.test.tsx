import { render, screen, waitFor } from '@testing-library/react';
import { beforeEach, describe, expect, test, vi } from 'vitest';
import SearchPage from './SearchPage';
import { MemoryRouter } from 'react-router';
import { QueryClient, QueryClientProvider } from '@tanstack/react-query';
import { searchHeroesAction } from '@/heores/actions/search-heroes.action';
import type { Hero } from '@/heores/types/hero.interface';

vi.mock('@/heores/actions/search-heroes.action');
const mockSearchHeroesAction = vi.mocked(searchHeroesAction);

vi.mock('@/components/custom/CustomJumbotron', () => ({
  CustomJumbotron: () => <div data-testid="custom-jumbotron"></div>,
}));

vi.mock('./ui/SearchControls', () => ({
  SearchControls: () => <div data-testid="search-controls"></div>
}));

vi.mock('@/heores/componentes/HeroGrid', () => ({
  HeroGrid: ({ heroes }: { heroes: Hero[] }) => (<div data-testid="hero-grid">
    {
      heroes.map(hero => (
        <div key={hero.id}>{hero.name}</div>
      ))
    }
  </div>)
}));

const queryClient = new QueryClient();

const renderSearchPage = (initialEntries: string[] = ['/']) => {
  return render(
    <MemoryRouter initialEntries={initialEntries}>
      <QueryClientProvider client={queryClient}>
        <SearchPage />
      </QueryClientProvider>
    </MemoryRouter>
  );
};

describe('SearchPage', () => {

  beforeEach(() => {
    vi.clearAllMocks();
  });

  test('should render SearchPage with default values', () => {
    const { container } = renderSearchPage();
    // screen.debug();

    expect(mockSearchHeroesAction).toHaveBeenCalledWith({
      name: undefined,
      strength: undefined,
    });

    expect(container).toMatchSnapshot();
  });

  test('should call search action with name parameter', () => {
    const { container } = renderSearchPage(['/search?name=superman']);

    expect(mockSearchHeroesAction).toHaveBeenCalledWith({
      name: 'superman',
      strength: undefined,
    });

    expect(container).toMatchSnapshot();
  });

  test('should call search action with strenght parameter', () => {
    const { container } = renderSearchPage(['/search?strength=7']);

    expect(mockSearchHeroesAction).toHaveBeenCalledWith({
      name: undefined,
      strength: '7',
    });

    expect(container).toMatchSnapshot();
  });

  test('should call search action with name and strenght parameters', () => {
    const { container } = renderSearchPage(['/search?name=batman&strength=8']);

    expect(mockSearchHeroesAction).toHaveBeenCalledWith({
      name: 'batman',
      strength: '8',
    });

    expect(container).toMatchSnapshot();
  });

  test('should render HeroGrid with search result', async () => {
    const mockHeroes = [
      { id: '1', name: 'Clark Kent' } as unknown as Hero,
      { id: '2', name: 'Bruce Wayne' } as unknown as Hero,
    ];

    mockSearchHeroesAction.mockResolvedValue(mockHeroes);

    renderSearchPage();

    //* Esperar a que esos elementos sean cargados
    await waitFor(() => {
      expect(screen.getByText('Clark Kent')).toBeDefined();
      expect(screen.getByText('Bruce Wayne')).toBeDefined();
    });

    screen.debug(screen.getByTestId('hero-grid'));
  });

});