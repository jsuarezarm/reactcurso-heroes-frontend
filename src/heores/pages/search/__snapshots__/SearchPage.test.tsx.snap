// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SearchPage > should call search action with name and strenght parameters 1`] = `
<div>
  <div
    data-testid="custom-jumbotron"
  />
  <nav
    aria-label="breadcrumb"
    class="my-5"
    data-slot="breadcrumb"
  >
    <ol
      class="text-muted-foreground flex flex-wrap items-center gap-1.5 text-sm break-words sm:gap-2.5"
      data-slot="breadcrumb-list"
    >
      <li
        class="inline-flex items-center gap-1.5"
        data-slot="breadcrumb-item"
      >
        <a
          class="hover:text-foreground transition-colors"
          data-slot="breadcrumb-link"
          href="/"
        >
          Inicio
        </a>
      </li>
      <div
        class="flex gap-2 items-center"
      >
        <li
          aria-hidden="true"
          class="[&>svg]:size-3.5"
          data-slot="breadcrumb-separator"
          role="presentation"
        >
          <svg
            aria-hidden="true"
            class="lucide lucide-slash"
            fill="none"
            height="24"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M22 2 2 22"
            />
          </svg>
        </li>
        <li
          class="inline-flex items-center gap-1.5"
          data-slot="breadcrumb-item"
        >
          <a
            data-discover="true"
            href="/search"
          >
            Search
          </a>
        </li>
      </div>
    </ol>
  </nav>
  <div>
    Cargando...
  </div>
  <div
    data-testid="search-controls"
  />
  <div
    data-testid="hero-grid"
  />
</div>
`;

exports[`SearchPage > should call search action with name parameter 1`] = `
<div>
  <div
    data-testid="custom-jumbotron"
  />
  <nav
    aria-label="breadcrumb"
    class="my-5"
    data-slot="breadcrumb"
  >
    <ol
      class="text-muted-foreground flex flex-wrap items-center gap-1.5 text-sm break-words sm:gap-2.5"
      data-slot="breadcrumb-list"
    >
      <li
        class="inline-flex items-center gap-1.5"
        data-slot="breadcrumb-item"
      >
        <a
          class="hover:text-foreground transition-colors"
          data-slot="breadcrumb-link"
          href="/"
        >
          Inicio
        </a>
      </li>
      <div
        class="flex gap-2 items-center"
      >
        <li
          aria-hidden="true"
          class="[&>svg]:size-3.5"
          data-slot="breadcrumb-separator"
          role="presentation"
        >
          <svg
            aria-hidden="true"
            class="lucide lucide-slash"
            fill="none"
            height="24"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M22 2 2 22"
            />
          </svg>
        </li>
        <li
          class="inline-flex items-center gap-1.5"
          data-slot="breadcrumb-item"
        >
          <a
            data-discover="true"
            href="/search"
          >
            Search
          </a>
        </li>
      </div>
    </ol>
  </nav>
  <div>
    Cargando...
  </div>
  <div
    data-testid="search-controls"
  />
  <div
    data-testid="hero-grid"
  />
</div>
`;

exports[`SearchPage > should call search action with strenght parameter 1`] = `
<div>
  <div
    data-testid="custom-jumbotron"
  />
  <nav
    aria-label="breadcrumb"
    class="my-5"
    data-slot="breadcrumb"
  >
    <ol
      class="text-muted-foreground flex flex-wrap items-center gap-1.5 text-sm break-words sm:gap-2.5"
      data-slot="breadcrumb-list"
    >
      <li
        class="inline-flex items-center gap-1.5"
        data-slot="breadcrumb-item"
      >
        <a
          class="hover:text-foreground transition-colors"
          data-slot="breadcrumb-link"
          href="/"
        >
          Inicio
        </a>
      </li>
      <div
        class="flex gap-2 items-center"
      >
        <li
          aria-hidden="true"
          class="[&>svg]:size-3.5"
          data-slot="breadcrumb-separator"
          role="presentation"
        >
          <svg
            aria-hidden="true"
            class="lucide lucide-slash"
            fill="none"
            height="24"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M22 2 2 22"
            />
          </svg>
        </li>
        <li
          class="inline-flex items-center gap-1.5"
          data-slot="breadcrumb-item"
        >
          <a
            data-discover="true"
            href="/search"
          >
            Search
          </a>
        </li>
      </div>
    </ol>
  </nav>
  <div>
    Cargando...
  </div>
  <div
    data-testid="search-controls"
  />
  <div
    data-testid="hero-grid"
  />
</div>
`;

exports[`SearchPage > should render SearchPage with default values 1`] = `
<div>
  <div
    data-testid="custom-jumbotron"
  />
  <nav
    aria-label="breadcrumb"
    class="my-5"
    data-slot="breadcrumb"
  >
    <ol
      class="text-muted-foreground flex flex-wrap items-center gap-1.5 text-sm break-words sm:gap-2.5"
      data-slot="breadcrumb-list"
    >
      <li
        class="inline-flex items-center gap-1.5"
        data-slot="breadcrumb-item"
      >
        <a
          class="hover:text-foreground transition-colors"
          data-slot="breadcrumb-link"
          href="/"
        >
          Inicio
        </a>
      </li>
    </ol>
  </nav>
  <div>
    Cargando...
  </div>
  <div
    data-testid="search-controls"
  />
  <div
    data-testid="hero-grid"
  />
</div>
`;
